components/mission_key_management/cn_stubs.h

```diff
@@ -4,77 +4,176 @@
 
 
 // Cerberus puts some POSIX headers under the `posix/` directory.
-#include <posix/sys/types.h>
-
+#include "policy.h"
+#include "client.h"
+
+// from `cn_memory.h`
+
+#if ! defined(CN_TEST)
+// For verification, use specs in ../../components/include/cn_memory.h
+# define memcpy(d,s,n) _memcpy(d,s,n)
+# define memcmp(s1,s2,n) _memcmp(s1,s2,n)
+# define malloc(x) _malloc_canfail(x)
+# define free(x) _free(x)
+#else 
+// For testing, use the CN-specific instrumented malloc() / free()
+void *cn_malloc(unsigned long size);
+void cn_free_sized(void *ptr, unsigned long size);
+# define malloc(s) cn_malloc(s)
+# define free(c) cn_free_sized(c, sizeof(struct client))
+#endif 
 
 // From `sys/epoll.h`
 #define EPOLLIN 1
 #define EPOLLOUT 4
 
+// From `sys/socket.h`
+#define SHUT_RDWR 2
+
+// From `policy.h`
+
+// Non-deterministically return a pointer to a key, or NULL 
+const uint8_t* _policy_match(uint8_t key_id[KEY_ID_SIZE], uint8_t nonce[NONCE_SIZE],
+        uint8_t measure[MEASURE_SIZE], uint8_t hmac[HMAC_SIZE]);
+/*$ spec _policy_match(pointer key_id, pointer nonce, pointer measure, pointer hmac); 
+requires 
+    true; 
+ensures 
+    take Key_out = KeyPred(return); 
+$*/
+#if ! defined(CN_TEST)
+#define policy_match(k,n,m,h) _policy_match(k,n,m,h) 
+#else 
+// Mock policy_match by allocating a new key 
+#define policy_match(...) cn_malloc(KEY_SIZE * sizeof(const uint8_t))
+#endif 
+
+// // Add an entry to the policy table.  Returns 0 on success and -1 on failure.
+// int _policy_add(
+//         const uint8_t key_id[KEY_ID_SIZE],
+//         const uint8_t measure[MEASURE_SIZE],
+//         const uint8_t key[KEY_SIZE]);
+// /*$ spec _policy_add(pointer key_id, pointer measure, pointer key); 
+// requires 
+//     true; 
+// ensures 
+//     return == 0i32 || return == -1i32; 
+// $*/
+// #define policy_add(k,m,h) _policy_add(k,m,h) 
+
+// Ghost function which releases the memory representing a key. Implicitly, this
+// is returning ownership of the memory to whatever internal state manages the
+// key list. 
+void _key_release (const uint8_t *key); 
+/*$ spec _key_release(pointer key); 
+requires 
+    take Key_in = KeyPred(key);
+ensures 
+    true; 
+$*/
+#if ! defined(CN_TEST) 
+#define key_release(k) _key_release(k)
+#else 
+// Mock key_release by disposing the key 
+#define key_release(k) if (k != NULL) { cn_free_sized((void *) k, KEY_SIZE * sizeof(const uint8_t)); } 
+#endif 
 
 // From `stdio.h`
 
-#ifndef WAR_CN_309
+/*$ spec fprintf(pointer f, pointer s);
+requires true;
+ensures true;
+$*/
+#if defined(CN_TEST)
+# define fprintf(...) 0 
```

Comment:
All these should defines really be in some sort of CN standard library...

Reply:
Yes, agreed. Really, files in `components/include` such as `cn_memory.h` are the beginning of such a standard library